ostd/src/logger.rs

```diff
@@ -10,96 +10,78 @@
 //!
 //! IRQs are disabled while printing. So do not print long log messages.
 
+use core::str::FromStr;
+
 use log::{LevelFilter, Metadata, Record};
+use spin::Once;
+
+static LOGGER: Logger = Logger::new();
 
-use crate::{
-    boot::{kcmdline::ModuleArg, kernel_cmdline},
-    timer::Jiffies,
-};
+struct Logger {
+    backend: Once<&'static dyn log::Log>,
+}
 
-const LOGGER: Logger = Logger {};
+impl Logger {
+    const fn new() -> Self {
+        Self {
+            backend: Once::new(),
+        }
+    }
+}
 
-struct Logger {}
+/// Injects a logger as the global logger backend.
+///
+/// This method allows upper-level users to inject their own implemented loggers,
+/// but only allows injecting once. Subsequent injection will have no effect.
+pub fn inject_logger(new_logger: &'static dyn log::Log) {
+    LOGGER.backend.call_once(|| new_logger);
+}
 
 impl log::Log for Logger {
     fn enabled(&self, metadata: &Metadata) -> bool {
-        metadata.level() <= log::max_level()
+        if let Some(logger) = self.backend.get() {
+            return logger.enabled(metadata);
+        };
+
+        // Default implementation.
+        true
     }
 
     fn log(&self, record: &Record) {
         if !self.enabled(record.metadata()) {
             return;
         }
 
-        let timestamp = Jiffies::elapsed().as_duration().as_secs_f64();
+        if let Some(logger) = self.backend.get() {
+            return logger.log(record);
+        };
+
+        // Default implementation.
```

Comment:
The default implementation is composed by multiple code blocks, so there should be a blank line before the "Default implementation" comment.